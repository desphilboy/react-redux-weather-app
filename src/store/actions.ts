import { createActions } from "redux-actions";

export const START_API = "START_API";
export const ERROR_API = "ERROR_API";
export const DONE_API = "DONE_API";

export const SET_COUNTRY = "SET_COUNTRY";

export const SET_UNITS = "SET_UNITS";

export const CLEAR_RECORDS = "CLEAR_RECORDS";

export const { startApi, doneApi, errorApi } = createActions({
	START_API: () => ({ data: false, inProgress: true, error: false }), // payload creator
	DONE_API: (data) => ({ data, inProgress: false, error: false }),
	ERROR_API: (error) => ({ data: false, inProgress: false, error }),
});

export const setUnits = (unitsType) => ({
	type: SET_UNITS,
	payload: unitsType,
});

export const setCountry = (country) => ({
	type: SET_COUNTRY,
	payload: country,
});

export const clearRecords = () => ({
	type: CLEAR_RECORDS,
});
