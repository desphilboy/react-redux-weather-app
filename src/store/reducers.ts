import {
   START_API,
   ERROR_API,
   DONE_API,
   SET_COUNTRY,
   SET_UNITS,
   CLEAR_RECORDS,
} from "./actions";

export const INITIAL_STATE = {
   country: "AU",
   units: "metric",

   apiCall: { data: false, error: false, inProgress: false },
   weatherRecords: [],
};

export const apiCall = (state = INITIAL_STATE.apiCall, action) => {
   switch (action.type) {
      case START_API:
      case ERROR_API:
      case DONE_API:
         return { ...state, ...action.payload };
      default:
         return state;
   }
};

export const country = (state = INITIAL_STATE.country, action) => {
   switch (action.type) {
      case SET_COUNTRY:
         return action.payload;
      default:
         return state;
   }
};

export const units = (state = INITIAL_STATE.units, action) => {
   switch (action.type) {
      case SET_UNITS:
         return action.payload;
      default:
         return state;
   }
};

export const weatherRecords = (
   state = INITIAL_STATE.weatherRecords,
   action
) => {
   switch (action.type) {
      case DONE_API:
         return [...state, action.payload.data];
      case CLEAR_RECORDS:
         return [];
      default:
         return state;
   }
};
