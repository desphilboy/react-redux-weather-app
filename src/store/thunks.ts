import { startApi, doneApi, errorApi } from "./actions";

const appId = "ce242e6f1909e3ca32395ead09817fb6";
const baseUrl = "http://api.openweathermap.org/data/2.5/weather";

export const fetchWeather = ({ city = "Melbourne" } = {}) => (
	dispatch,
	getState
) => {
	const { country, units } = getState();

	let fetchurl = `${baseUrl}?q=${city || "Melbourne"},${
		country || "AU"
	}&units=${units || "metric"}&appid=${appId}`;

	dispatch(startApi());
	fetch(fetchurl)
		.then((response) => response.json())
		.then((data) => {
			console.log("data in fetchWeather", data);
			dispatch(doneApi(data));
		})
		.catch((err) => {
			console.log(err);
			dispatch(errorApi(err));
		});
};
