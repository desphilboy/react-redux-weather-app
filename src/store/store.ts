import { createStore, applyMiddleware, combineReducers } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { apiCall, country, units } from "./reducers";

const reducer = combineReducers({ apiCall, country, units });

const store = createStore(reducer, composeWithDevTools(applyMiddleware(thunk)));

export default store;
