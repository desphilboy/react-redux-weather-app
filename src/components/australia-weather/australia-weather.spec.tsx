import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import store from "../../store/store.ts";
import Component, { AustraliaWeather } from "./austalia-weather";

describe("austaliaWeather Serach component", () => {
	const changeCountryMock = jest.fn();
	const fetchConditionsMock = jest.fn();

	describe("View", () => {
		beforeEach(() => {
			jest.clearAllMocks();
		});

		it("matches the snapshot", () => {
			const wrapper = render(
				<AustraliaWeather
					changeCountry={changeCountryMock}
					fetchConditions={fetchConditionsMock}
				/>
			);
			expect(wrapper).toMatchSnapshot();
		});

		it("renders a label", async () => {
			const { findAllByTestId } = render(
				<AustraliaWeather
					changeCountry={changeCountryMock}
					fetchConditions={fetchConditionsMock}
				/>
			);
			expect(
				await findAllByTestId("select-city-label")
			).toMatchSnapshot();
			expect(await findAllByTestId("city-name-input")).toMatchSnapshot();
		});

		it("renders an input for city name", async () => {
			const { findAllByTestId } = render(
				<AustraliaWeather
					changeCountry={changeCountryMock}
					fetchConditions={fetchConditionsMock}
				/>
			);
			expect(await findAllByTestId("city-name-input")).toMatchSnapshot();
		});

		it("renders an input for city name", async () => {
			const { findAllByTestId } = render(
				<AustraliaWeather
					changeCountry={changeCountryMock}
					fetchConditions={fetchConditionsMock}
				/>
			);
			expect(await findAllByTestId("city-name-input")).toMatchSnapshot();
		});
	});

	describe("Rendering with Redux Store", () => {
		it("renders with redux store", () => {
			const wrapper = render(
				<Provider store={store}>
					<Component />
				</Provider>
			);
			expect(wrapper).toMatchSnapshot();
		});

		it("typing a city name and prssing enter", async () => {
			let action;
			store.dispatch = jest.fn((param) => (action = param));
			const { getByTestId } = render(
				<Provider store={store}>
					<Component />
				</Provider>
			);
			const cityNameInput = getByTestId("city-name-input");
			fireEvent.keyPress(cityNameInput, { key: "A", code: "keyA" });
			fireEvent.keyDown(cityNameInput, { key: "Enter", code: "Enter" });

			expect(store.dispatch).toHaveBeenCalledWith(expect.any(Function));
		});
	});
});