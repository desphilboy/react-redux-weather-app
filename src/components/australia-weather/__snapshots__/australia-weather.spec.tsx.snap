// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`austaliaWeather Serach component Rendering with Redux Store renders with redux store 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdnylx jWKbmi"
      >
        <div
          class="sc-gtssRu VrVmy"
        >
          <label
            data-testid="select-city-label"
          >
            Select city from list:
          </label>
          <select
            name="city"
          >
            <option
              value="Sydney"
            >
              Sydney
            </option>
            <option
              value="Melbourne"
            >
              Melbourne
            </option>
            <option
              value="Alice Springs"
            >
              Alice Springs
            </option>
            <option
              value="Adelaide"
            >
              Adelaide
            </option>
            <option
              value="Hobart"
            >
              Hobart
            </option>
            <option
              value="Perth"
            >
              Perth
            </option>
            <option
              value="Canberra"
            >
              Canberra
            </option>
            <option
              value="Brisbane"
            >
              Brisbane
            </option>
          </select>
        </div>
        <div
          class="sc-gtssRu VrVmy"
        >
          <label>
             or type city's name:
          </label>
          <input
            data-testid="city-name-input"
            name="city"
            type="text"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnylx jWKbmi"
    >
      <div
        class="sc-gtssRu VrVmy"
      >
        <label
          data-testid="select-city-label"
        >
          Select city from list:
        </label>
        <select
          name="city"
        >
          <option
            value="Sydney"
          >
            Sydney
          </option>
          <option
            value="Melbourne"
          >
            Melbourne
          </option>
          <option
            value="Alice Springs"
          >
            Alice Springs
          </option>
          <option
            value="Adelaide"
          >
            Adelaide
          </option>
          <option
            value="Hobart"
          >
            Hobart
          </option>
          <option
            value="Perth"
          >
            Perth
          </option>
          <option
            value="Canberra"
          >
            Canberra
          </option>
          <option
            value="Brisbane"
          >
            Brisbane
          </option>
        </select>
      </div>
      <div
        class="sc-gtssRu VrVmy"
      >
        <label>
           or type city's name:
        </label>
        <input
          data-testid="city-name-input"
          name="city"
          type="text"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`austaliaWeather Serach component View matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdnylx jWKbmi"
      >
        <div
          class="sc-gtssRu VrVmy"
        >
          <label
            data-testid="select-city-label"
          >
            Select city from list:
          </label>
          <select
            name="city"
          >
            <option
              value="Sydney"
            >
              Sydney
            </option>
            <option
              value="Melbourne"
            >
              Melbourne
            </option>
            <option
              value="Alice Springs"
            >
              Alice Springs
            </option>
            <option
              value="Adelaide"
            >
              Adelaide
            </option>
            <option
              value="Hobart"
            >
              Hobart
            </option>
            <option
              value="Perth"
            >
              Perth
            </option>
            <option
              value="Canberra"
            >
              Canberra
            </option>
            <option
              value="Brisbane"
            >
              Brisbane
            </option>
          </select>
        </div>
        <div
          class="sc-gtssRu VrVmy"
        >
          <label>
             or type city's name:
          </label>
          <input
            data-testid="city-name-input"
            name="city"
            type="text"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnylx jWKbmi"
    >
      <div
        class="sc-gtssRu VrVmy"
      >
        <label
          data-testid="select-city-label"
        >
          Select city from list:
        </label>
        <select
          name="city"
        >
          <option
            value="Sydney"
          >
            Sydney
          </option>
          <option
            value="Melbourne"
          >
            Melbourne
          </option>
          <option
            value="Alice Springs"
          >
            Alice Springs
          </option>
          <option
            value="Adelaide"
          >
            Adelaide
          </option>
          <option
            value="Hobart"
          >
            Hobart
          </option>
          <option
            value="Perth"
          >
            Perth
          </option>
          <option
            value="Canberra"
          >
            Canberra
          </option>
          <option
            value="Brisbane"
          >
            Brisbane
          </option>
        </select>
      </div>
      <div
        class="sc-gtssRu VrVmy"
      >
        <label>
           or type city's name:
        </label>
        <input
          data-testid="city-name-input"
          name="city"
          type="text"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`austaliaWeather Serach component View renders a label 1`] = `
Array [
  <label
    data-testid="select-city-label"
  >
    Select city from list:
  </label>,
]
`;

exports[`austaliaWeather Serach component View renders a label 2`] = `
Array [
  <input
    data-testid="city-name-input"
    name="city"
    type="text"
  />,
]
`;

exports[`austaliaWeather Serach component View renders an input for city name 1`] = `
Array [
  <input
    data-testid="city-name-input"
    name="city"
    type="text"
  />,
]
`;

exports[`austaliaWeather Serach component View renders an input for city name 2`] = `
Array [
  <input
    data-testid="city-name-input"
    name="city"
    type="text"
  />,
]
`;
